### PR TITLE
Rebot connection timeout

### PR DESCRIPTION
**Pull request by vargheseg**
Pull request brings in changes for ticket https://github.com/ChimeraTK/DeviceAccess/issues/56

### CODE REVIEW

File: tests/unitTestsNotUnderCtest/rebotManualTest.cpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
+#include "Device.h"
```
Comment (killenb) [inline tests/unitTestsNotUnderCtest/rebotManualTest.cpp]: Very nice :-)

File: device_backends/Rebot/include/RebotBackend.h [hunk 2]
```diff
@@ -40,40 +43,42 @@ namespace ChimeraTK {
...
-    // This is not in the protocol implementor. Only the result of the hello tells
-    // us which implementor to instantiate.
-    uint32_t getServerProtocolVersion();
-    std::vector<uint32_t> frameClientHello();
-    uint32_t parseRxServerHello(const std::vector<int32_t>& serverHello);
 
     void heartbeatLoop(boost::shared_ptr<ThreadInformerMutex> threadInformerMutex);
     boost::thread _heartbeatThread;
+
+    const static uint32_t CONNECTION_TIMEOUT_SEC{5};
```
[the remark this answers is not in the document]
Comment (vargheseg) [reply device_backends/Rebot/include/RebotBackend.h]: Yes it is. 

The default is used if the user does not have a  timeout in the ChimeraTK ccd.

[the remark this answers is not in the document]
Comment (vargheseg) [reply device_backends/Rebot/include/RebotBackend.h]: will revise  name to DEFAULT_CONNECTION_TIMEOUT_SEC

File: device_backends/Rebot/src/RebotBackend.cc [hunk 3]
```diff
@@ -1,53 +1,103 @@
...
+    _threadInformerMutex->quitThread = true;
 
-      if(isOpen()) {
-        _tcpCommunicator->closeConnection();
-      }
-    } // end of the lock guard scope. We have to release the lock before waiting
-      // for the thread to join
-    _heartbeatThread.interrupt();
-    _heartbeatThread.join();
+    if (isOpen()) {
```
[the remark this answers is not in the document]
Comment (vargheseg) [reply device_backends/Rebot/src/RebotBackend.cc]: This is supported; consecutive calls to  RebotBackend::open are permitted

File: device_backends/Rebot/src/RebotProtocol0.cc [hunk 4]
```diff
@@ -48,67 +48,46 @@ namespace ChimeraTK {
...
-      }
-      for(int j = 8; j < 12; ++j) {
-        datasend[j] = (data[i] >> (8 * (j - 8))) & 0xFF;
-      }
-
-      boost::array<char, 4> receivedData;
-      _tcpCommunicator->sendData(datasend);
-      _tcpCommunicator->receiveData(receivedData);
-      // FIXME: Do error handling of the response
+        packet.at(0) = SINGLE_WORD_WRITE;
```
[the remark this answers is not in the document]
Comment (vargheseg) [reply device_backends/Rebot/src/RebotProtocol0.cc]: we now have le32toh and htole32  on read and written application data

File: device_backends/Rebot/src/RebotProtocol0.cc [hunk 5]
```diff
@@ -48,67 +48,46 @@ namespace ChimeraTK {
...
+        packet.at(1) = registerInfo.addressInWords++;
+        packet.at(2) = static_cast<uint32_t>(data[i]);
+        _tcpCommunicator->write(packet);
+        _tcpCommunicator->read(1); // response is ignored for now
```
[the remark this answers is not in the document]
Comment (vargheseg) [reply device_backends/Rebot/src/RebotProtocol0.cc]: This is to do with how the protocol was specified initially; return values have been broken from day0 and not fixed.

File: device_backends/src/BackendFactory.cc [hunk 6]
```diff
@@ -140,7 +140,7 @@ namespace ChimeraTK {
     registerBackendType("pci", &PcieBackend::createInstance, {"map"});
 #endif
     registerBackendType("dummy", &DummyBackend::createInstance, {"map"});
-    registerBackendType("rebot", &RebotBackend::createInstance, {"ip", "port", "map"});
+    registerBackendType("rebot", &RebotBackend::createInstance, {"ip", "port", "map", "timeout"});
```
[the remark this answers is not in the document]
Comment (vargheseg) [reply device_backends/src/BackendFactory.cc]: timeout is configurable

File: device_backends/Rebot/src/RebotProtocol0.cc [hunk 7]
```diff
@@ -48,67 +48,46 @@ namespace ChimeraTK {
...
-    }
-    for(int j = 4; j < 8; ++j) {
-      datasend[j] = (wordAddress >> (8 * (j - 4))) & 0xFF;
-    }
-    for(int j = 8; j < 12; ++j) {
-      datasend[j] = ((wordsToRead) >> (8 * (j - 8))) & 0xFF;
-    }
-
-    _tcpCommunicator->sendData(datasend);
+    std::vector<uint32_t> packet(3);
```
[the remark this answers is not in the document]
Comment (vargheseg) [reply device_backends/Rebot/src/RebotProtocol0.cc]: nope;

File: device_backends/Rebot/src/RebotProtocol0.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (killenb) [reply device_backends/Rebot/src/RebotProtocol0.cc]: We decided do remove the big-endian compatibility attempt. It never was complete and we cannot test it anyway. We have no use case for it now. We can come back to it when we need it.

**Pull request merged**